soci/soci_convert.go

```diff
@@ -0,0 +1,310 @@
+		if err != nil {
+			return nil, err
+		}
+	}
+	if convertCfg.gcRoot {
+		err := store.LabelGCRoot(ctx, b.blobStore, ociIndexDesc)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return &ociIndexDesc, nil
+}
+
+// buildPlatforms builds a SOCI index for each specified platform
+func (b *IndexBuilder) buildPlatforms(ctx context.Context, img images.Image, platforms []ocispec.Platform) ([]*IndexWithMetadata, error) {
+	var indexes []*IndexWithMetadata
+	for _, platform := range platforms {
+		index, err := b.Build(ctx, img,
+			WithPlatform(platform),
+			WithNoGarbageCollectionLabel(),
+			withIndexVersion(V2),
+		)
+		if err != nil {
+			// Skipping a platform is fine
+			if errors.Is(err, ErrEmptyIndex) {
+				continue
+			}
+			return nil, err
+		}
+		indexes = append(indexes, index)
+	}
+
+	if len(indexes) == 0 {
+		return nil, ErrEmptyIndex
+	}
+	return indexes, nil
+}
+
+// newOciIndex creates an OCI index object for the converted image.
+// If `img.Target` is already an OCI index (or docker manifest list), newOciIndex loads that index,
+// otherwise, it creates a new OCI index containing just the single image target.
+func (b *IndexBuilder) newOciIndex(ctx context.Context, img images.Image) (ocispec.Index, error) {
+	if images.IsIndexType(img.Target.MediaType) {
+		// Load the original target
+		ra, err := b.contentStore.ReaderAt(ctx, img.Target)
+		if err != nil {
+			return ocispec.Index{}, err
+		}
+		b, err := io.ReadAll(io.NewSectionReader(ra, 0, ra.Size()))
+		if err != nil {
+			return ocispec.Index{}, err
+		}
+		err = ociutil.ValidateMediaType(b, img.Target.MediaType)
+		if err != nil {
+			return ocispec.Index{}, err
+		}
+
+		var ociIndex ocispec.Index
+		err = json.Unmarshal(b, &ociIndex)
+		// Some Registries don't like it when you push a Docker v2 manifest
+		// To increase comptatibility, change the media type to OCI image manifest
+		ociIndex.MediaType = ocispec.MediaTypeImageIndex
+		return ociIndex, err
+	}
+	return ocispec.Index{
+		Versioned: specs.Versioned{
+			SchemaVersion: 2,
+		},
+		MediaType: ocispec.MediaTypeImageIndex,
+		Manifests: []ocispec.Descriptor{
+			img.Target,
+		},
+	}, nil
+}
+
+// annotateImages adds the SOCI index digest to the corresponding image manifest descriptor, modifying `ociIndex` in the process.
+// Adding annotations modifies the image manifest, so annotateImages also pushes the modified image manifests to the blobStore,
+// computes the new image digests, and modified the image descriptors in `ociIndex`
+func (b *IndexBuilder) annotateImages(ctx context.Context, ociIndex *ocispec.Index, sociIndexes []*IndexWithMetadata) error {
+	for i := 0; i < len(ociIndex.Manifests); i++ {
+		manifestDesc := &ociIndex.Manifests[i]
+		// images.Manifest validates the mediatype, no need to do it ourselves like
+		// we did when loading the OCI index
+		manifest, err := images.Manifest(ctx, b.contentStore, *manifestDesc, nil)
```

Comment:
qq: does this return the original manifest? In `newOciIndex`, we copied the contents of the original oci index, but the manifests should still refer to the same digest, right? So if we modify the manifest here, we no longer have access to the original manifest.

Reply:
Yes, this is intentional. We load the original image manifest, add the annotation, serialize the new manifest and overwrite the digest in the OCI index.